File: 821.字符的最短距离.py
#
# @lc app=leetcode.cn id=821 lang=python3
#
# [821] 字符的最短距离
#

# @lc code=start
class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        m = len(s)
        def getDistance(index: int) -> int:
            if s[index] == c:
                return 0
            left, right = index, index
            while left >= 0 or right < m:
                if left >= 0 and s[left] == c:
                    return abs(index - left)
                elif left >= 0:
                    left -= 1
                if right < m and s[right] == c:
                    return abs(index - right)
                elif right < m:
                    right += 1
            return -1
        res = [getDistance(i) for i in range(m)]
        return res
# @lc code=end

File: 821.字符的最短距离.go
/*
 * @lc app=leetcode.cn id=821 lang=golang
 *
 * [821] 字符的最短距离
 */

// @lc code=start
// 执行用时：0 ms, 在所有 Go 提交中击败了100.00%的用户
// 内存消耗：2.2 MB, 在所有 Go 提交中击败了100.00%的用户
func shortestToChar(s string, c byte) []int {
    m := len(s)
    var getDistance func(index int) int
    getDistance = func(index int) (res int) {
        if s[index] == c {
            return
        }
        left, right := index, index
        for left >= 0 || right < m {
            if left >= 0 {
                if s[left] == c {
                    res = abs(index - left)
                    return
                }
                left--
            }
            if right < m {
                if s[right] == c {
                    res = abs(index - right)
                    return
                }
                right++
            }
        }
        return
    }
    res := make([]int, m)
    for i := 0; i < m; i++ {
        res[i] = getDistance(i)
    }
    return res
}

func abs(num int) int {
    if num < 0 {
        return -num
    }
    return num
}
// @lc code=end

